### PR TITLE
token group: init interface

### PR DESCRIPTION
**Pull request by buffalojoec**
This PR initializes a newer, slimmed down version of the SPL Token Group interface.

Previous PR: #5190

### CODE REVIEW

File: token-group/interface/src/instruction.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+/// Instruction data for initializing a new `Member` of a `Group`
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:initialize_member")]
+pub struct InitializeMember {
+    /// The pubkey of the `Group`
+    pub group: Pubkey,
+    /// The member number
+    pub member_number: u32,
```
Comment (joncinque) [inline token-group/interface/src/instruction.rs:54]: Why is this needed? Seems like it should be able to figure it out from the group, no?

File: token-group/interface/src/instruction.rs [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    pub new_authority: OptionalNonZeroPubkey,
+}
+
+/// Instruction data for initializing a new `Member` of a `Group`
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:initialize_member")]
+pub struct InitializeMember {
+    /// The pubkey of the `Group`
+    pub group: Pubkey,
```
Comment (joncinque) [inline token-group/interface/src/instruction.rs:52]: Why is this needed? It's already provided as an account

File: token-group/interface/src/instruction.rs [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    UpdateGroupAuthority(UpdateGroupAuthority),
+
+    /// Initialize a new `Member` of a `Group`
+    ///
+    /// Assumes the `Group` has already been initialized,
+    /// as well as the mint for the member.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[w]`  Member
```
Comment (joncinque) [inline token-group/interface/src/instruction.rs:100]: Interesting, if I'm reading this right, since this can just be any account, then a mint can be a member of multiple groups. Is that correct?

File: token-group/interface/src/instruction.rs [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    ///   1. `[]`   Member Mint
+    ///   2. `[s]`  Member Mint authority
+    ///   3. `[w]`  Group
+    ///   4. `[]`   Group Mint
+    ///   5. `[s]`  Group Mint authority
```
Comment (joncinque) [inline token-group/interface/src/instruction.rs:105]: I'm thinking this should be the group update authority since the group exists. What do you think?

File: token-group/interface/src/instruction.rs [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+    /// Assumes the `Group` has already been initialized,
+    /// as well as the mint for the member.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[w]`  Member
+    ///   1. `[]`   Member Mint
+    ///   2. `[s]`  Member Mint authority
+    ///   3. `[w]`  Group
+    ///   4. `[]`   Group Mint
```
Comment (joncinque) [inline token-group/interface/src/instruction.rs:104]: What will this be needed for?

**Review by joncinque — COMMENTED** [review]
This is looking really good! Just some bits to iron out on the interface, then this should be good to start on

File: token-group/interface/src/instruction.rs (hunk 3 above)
Comment (buffalojoec) [reply token-group/interface/src/instruction.rs:100]: Yep, if that's what we want! Right now I left it up for discussion.

File: token-group/interface/src/instruction.rs (hunk 4 above)
Comment (buffalojoec) [reply token-group/interface/src/instruction.rs:105]: I could go either way tbh. I think update authority makes more since only because you might have an NFT as the group, and the mint authority is `None`, so that's prob preferred.

Note I was also considering removing the member mint authority, wdyt?

File: token-group/interface/src/instruction.rs (hunk 5 above)
Comment (buffalojoec) [reply token-group/interface/src/instruction.rs:104]: So no group mint then? I'm sure it doesn't _need_ to be part of the interface yeah.

File: token-group/interface/src/instruction.rs (hunk 2 above)
Comment (buffalojoec) [reply token-group/interface/src/instruction.rs:52]: Good point ha

File: token-group/interface/src/instruction.rs [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    /// The member number
+    pub member_number: u32,
+}
+
+/// All instructions that must be implemented in the SPL Token Group Interface
+///
+/// Note: Any instruction can be extended using additional required accounts by
+/// using the `InitializeExtraAccountMetaList` instruction to write
+/// configurations for extra required accounts into validation data
+/// corresponding to an instruction's unique discriminator.
```
[the remark this answers is not in the document]
Comment (buffalojoec) [reply token-group/interface/src/instruction.rs:62]: This snuck in there 🫠

**Comment by buffalojoec** [comment]
OK I've resolved most of the comments, but there's still a few items to discuss:
- The `u32` in place of `u64` for things like `max_size` is a result of #5518 
- Multiple groups for one member
- The instructions discriminators being a high number

File: token-group/interface/src/instruction.rs [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+    spl_pod::{
+        bytemuck::{pod_bytes_of, pod_from_bytes},
+        optional_keys::OptionalNonZeroPubkey,
+    },
+};
+
+/// Instruction data for initializing a new `Group`
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:initialize_group")]
```
[the remark this answers is not in the document]
Comment (joncinque) [reply token-group/interface/src/instruction.rs:20]: Correct, I just check for a first number that's "pretty big", at least in the 100s

File: token-group/interface/src/instruction.rs (hunk 4 above)
Comment (joncinque) [reply token-group/interface/src/instruction.rs]: Hm, I'd say let's keep the member mint authority, because then someone else can include your mint in all sorts of groups without your permission, which would probably help confuse people a lot.

File: token-group/interface/src/instruction.rs (hunk 5 above)
Comment (joncinque) [reply token-group/interface/src/instruction.rs]: Yeah correct, I don't think we need any information from the mint

File: token-group/interface/src/instruction.rs (hunk 3 above)
Comment (joncinque) [reply token-group/interface/src/instruction.rs:90]: awesome, that works!

File: token-group/interface/src/instruction.rs [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+pub struct InitializeGroup {
+    /// Update authority for the group
+    pub update_authority: OptionalNonZeroPubkey,
+    /// The maximum number of group members
+    pub max_size: u32,
```
Comment (joncinque) [inline token-group/interface/src/instruction.rs:25]: You'll likely need to use `PodU32` for the instructions and structs to avoid padding / alignment issues

File: token-group/interface/src/instruction.rs [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+    }
+}
+
+/// Creates a `InitializeMember` instruction
+#[allow(clippy::too_many_arguments)]
+pub fn initialize_member(
+    program_id: &Pubkey,
+    group: &Pubkey,
+    group_update_authority: &Pubkey,
+    member: &Pubkey,
```
Comment (joncinque) [inline token-group/interface/src/instruction.rs:227]: micro-nit: I like to have the params match the instruction
```suggestion
    member: &Pubkey,
    group: &Pubkey,
    group_update_authority: &Pubkey,
```

File: token-group/interface/src/state.rs [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq, Pod, Zeroable, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:group")]
+pub struct Group {
+    /// The authority that can sign to update the group
+    pub update_authority: OptionalNonZeroPubkey,
+    /// The current number of group members
+    pub size: u32,
+    /// The maximum number of group members
+    pub max_size: u32,
```
Comment (joncinque) [inline token-group/interface/src/state.rs:21]: Same here, these should probably be `PodU32`s

File: token-group/interface/src/instruction.rs [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+    pub max_size: u32,
+}
+
+/// Instruction data for updating the max size of a `Group`
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:update_group_max_size")]
+pub struct UpdateGroupMaxSize {
+    /// New max size for the group
+    pub max_size: u32,
```
Comment (joncinque) [inline token-group/interface/src/instruction.rs:34]: `PodU32` here

**Review by joncinque — COMMENTED** [review]
Just a few last things, then this should be good to go!

**Comment by buffalojoec** [comment]
> Edit: oh, and we'll also need the member mint authority to sign the `initialize_member` instruction

Would you say mint authority or update authority? A case could be made for either imo.

**Comment by buffalojoec** [comment]
Also, I think we're good on discriminators:

- `InitializeGroup`: [88, 13, 213, 85, 206, 182, 83, 46]
- `UpdateGroupAuthority`: [67, 150, 244, 200, 81, 37, 35, 55]
- `UpdateGroupMaxSize`: [108, 37, 171, 143, 248, 30, 18, 110]
- `InitializeMember`: [152, 32, 222, 176, 223, 237, 116, 134]

**Comment by buffalojoec** [comment]
Current dependencies on/for this PR:
* master
  * **PR #5512** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-program-library/5512" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-program-library/5512?utm_source=stack-comment).

**Comment by joncinque** [comment]
> Would you say mint authority or update authority? A case could be made for either imo.

There is no update authority here since we're just initializing the member, right? I think that means we have to go with member mint authority

**Review by joncinque — APPROVED** [review]
Looks good!

File: token-group/interface/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+version = "0.1.0"
+description = "Solana Program Library Token Group Interface"
+authors = ["Solana Labs Maintainers <maintainers@solanalabs.com>"]
+repository = "https://github.com/solana-labs/solana-program-library"
+license = "Apache-2.0"
+edition = "2021"
+
+[dependencies]
+bytemuck = "1.14.0"
+solana-program = "1.16.3"
```
Comment (joncinque) [inline token-group/interface/Cargo.toml:12]: Just need to get this correct so that the automatic updater works later
```suggestion
solana-program = "1.16.16"
```

**Review by joncinque — APPROVED** [review]
Looks great!

**Comment by buffalojoec** [comment]
Sorry @joncinque! Had to fix a `rustfmt` error!

**Review by joncinque — APPROVED** [review]
No worries! Feel free to merge without an explicit re-approval if you just fixed a rustfmt issue
